Tutorial - How to set up Next.js analytics, feature flags, and more

ivanagas opened the pull request:
## Changes

Mega Next.js tutorial: how to set up a basic blog app with authentication, then add PostHog, and set up custom event capture (including pageviews), user identification, and feature flags (client and server)

Hopefully solves some of the issues in: https://github.com/PostHog/posthog-js/issues/492

## Checklist
- [x] Titles are in [sentence case](https://apastyle.apa.org/style-grammar-guidelines/capitalization/sentence-case)
- [x] Feature names are in **[sentence case too]([https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case](https://apastyle.apa.org/style-grammar-guidelines/capitalization/sentence-case))**
- [x] Words are spelled using American English
- [x] I have checked out our [style guide](https://github.com/PostHog/posthog.com/blob/master/STYLEGUIDE.md)

joethreepwood commented on contents/tutorials/nextjs-analytics.md:
```diff
@@ -0,0 +1,671 @@
+2. pass the ID as a string to `getStaticProps()` to get the right blog for the route
+3. pass the post data to the component and render the data in HTML
+
+This looks like this:
```
```suggestion
Like this:
```

joethreepwood commented on contents/tutorials/nextjs-analytics.md:
```diff
@@ -0,0 +1,671 @@
+
+Next, add the session details, and the ability to sign in and out to our `index.js` page. We can do this with the `next-auth` `useSession()` hook. `next-auth` also provides methods for signing in and out. 
+
+ If there is a session, show details about the user and a button to sign out. If there isn’t a session, show a button to sign in. Together, this looks like this:
```
```suggestion
If there is a session, show details about the user and a button to sign out. If there isn’t a session, show a button to sign in. Together, it looks like this:
```

joethreepwood commented on contents/tutorials/nextjs-analytics.md:
```diff
@@ -0,0 +1,671 @@
+3. when the feature flags load, check the `blog-cta` flag we created earlier
+4. set the state for the CTA component
+
+This looks like this:
```
```suggestion
Like this:
```

charlescook-ph commented:
I'm going through this tutorial right now and have had a bunch of issues btw (99% of which I'm sure are because I'm a n00b but will write them all up just in case...)

joethreepwood commented:
> I'm going through this tutorial right now and have had a bunch of issues btw (99% of which I'm sure are because I'm a n00b but will write them all up just in case...)

This makes me wonder if we need to tag or sort tutorials somehow -- like the idea of having 'tutorial tracks' which we discussed at the offsite.

charlescook-ph commented:
This tutorial looked fun so I tried to run through it all myself! Unfortunately I wasn't successful - not sure if that is the tutorial's fault or because it's simply not aimed at a non-technical person like me? 

Anyway, here are my full notes - I spent 1.5hrs trying to get it to work:

- The image doesn't look quite right in preview, and there's no reading time or difficulty level included like we have with others? Maybe prerequisites would be helpful (e.g. this assumes you know JS or whatever - I assume I probably didn't meet the prerequisites for this one!)
- Link to Node latest version? (Does it matter which version I'm on?)
- After running `npx create-next-app@latest` I just hit 'y' here:

```
Need to install the following packages:

  create-next-app@13.1.6

Ok to proceed? (y) 
```
- I hit default for all the options, when it came to What import alias would you like configured? I hit enter again which was counterintuitive but worked.
- I didn't immediately understand that the name of the project = the name of the folder to navigate to (ie. I didn't realize you had called your project `tutorial`)
- In the _Adding blog functionality_ section I initially didn't realize I'd need to step out, create a JSON file in VSCode then come back to the Next app (and I had to poke around to find the `index.tsx` folder, not `index.js` right? `index.js` didn't do anything). I didn't realize that 'remove all the boilerplate' meant just delete all the text in those files but leave the files themselves intact. 
- Didn't realize at first that the `getstaticprops` method was already provided, I thought the instruction was to figure it out myself. 
- Worth adding a note why we are doing the authentication bit? Seemed a bit random to me to have a whole section on that (and I got really stuck/didn't understand the point of it). 
- At the _installing NextAuth_ part, for some reason my Terminal stopped 'responding' (presumably because I did something dumb). I just opened a new Terminal tab. 
- I assumed that I should just copy the GitHub details in the screenshot?
- Where is the `.env.local` file supposed to go? I just put it in the `auth` folder. Got some alert about using a period at the beginning which I ignored.
- Under _setting up sessions_, it took me aaages to find the `index.js` and `app.js` files
- I couldn't get this bit to work - the sign out button never showed up for me. I suspect I was putting the wrong things in the `index.js` vs. `index. tsx` file earlier so I eventually gave up once I could get my page to show up properly again (minus sign out button). 
- 'At this point, you need a PostHog instance.' - as in I should sign up for PostHog Cloud, or the hobbyist version? I went for Cloud anyway. 
- The script snippet link completely threw me because the page you link to talks about <head> tags? The whole flow here really confused me - is it reasonable to assume that the person already has a PostHog instance, or that they will go set one up for the first time now?
- In the end I pasted the code sample you included into `_app.js` and changed out the PostHog snippet bit for the one in my project settings, ignoring all the Head tag stuff. 
- I messed around clicking in http://localhost:3000/ at this point trying to get events to show up. It didn't work. Maybe because I didn't do the auth step properly? 
- I gave up at this point. 

Fully appreciate I may not be the target audience for this in which case feel free to disregard basically all this feedback? But I was bummed I couldn't get it to work, albeit mildly inspired to try and get into Next.js at least...

ivanagas commented:
All relevant feedback @charlescook-ph, I'm sorry you couldn't figure it out. Ideally, the tutorial would work for a motivated non-technical person. 

A classic hangup for non-technical people getting into coding is setting up the environment and I didn't cover it well enough. It was unclear whether to use TypeScript or not (Joe also made a comment), I wrote the tutorial **not using** TypeScript, and it looks like you did it **using** TypeScript. This caused some issues and changed file names. I updated to try to make clearer.

I've tried to address all of your other feedback as well from `getstaticprops` to why we need authorization to file locations to the PostHog snippet.

ivanagas commented on contents/tutorials/nextjs-analytics.md:
````diff
@@ -0,0 +1,671 @@
+npx create-next-app@latest
+```
+
+Name your app and select the defaults for everything (this tutorial won’t use TypeScript though). Once installed and created, go into the new folder (mine is `tutorial`) and start the server:
````
I've rewritten to clarify. This tutorial doesn't use TypeScript and using it will mess everything up.

charlescook-ph commented:
Sweet thanks! I'm going to take another crack at this tomorrow 😄

weyert commented:
Is there a particular reason, why the `posthog-js/react` wasn't used?

ivanagas commented:
> Is there a particular reason, why the `posthog-js/react` wasn't used?

There are some pieces of the posthog-js library that don't work well in next, see: https://github.com/PostHog/posthog-js/issues/492

Also we are working on an update to the react library: https://github.com/PostHog/posthog-js/pull/540
